modin/core/dataframe/base/dataframe/dataframe.py

```diff
@@ -0,0 +1,506 @@
+
+        Parameters
+        ----------
+        column_labels : string or list of strings
+            Column label(s) to use as the new row labels.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe with the row labels replaced by the specified columns.
+
+        Notes
+        -----
+        When multiple column labels are specified, a heirarchical set of labels is created, ordered by the ordering
+            of labels in the input.
+        """
+        pass
+
+    @abstractmethod
+    def from_labels(self) -> "ModinDataframe":
+        """
+        Move the row labels into the data at position 0, and sets the row labels to the positional notation.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe with the row labels moved into the data.
+
+        Notes
+        -----
+        In the case that the dataframe has hierarchical labels, all label "levels” are inserted into the dataframe
+            in the order they occur in the labels, with the outermost being in position 0.
+        """
+        pass
+
+    @abstractmethod
+    def rename(
+        self,
+        new_row_labels: Optional[Union[Dict[Hashable, Hashable], Callable]] = None,
+        new_col_labels: Optional[Union[Dict[Hashable, Hashable], Callable]] = None,
+        level: Optional[Union[int, List[int]]] = None,
+    ) -> "ModinDataframe":
+        """
+        Replace the row and column labels with the specified new labels.
+
+        Parameters
+        ----------
+        new_row_labels : dictionary or callable, default: None
+            Mapping from old row labels to new labels.
+        new_col_labels : dictionary or callable, default: None
+            Mapping from old col labels to new labels.
+        level : int or list of ints, default: None
+            Level(s) whose row labels to replace.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe with the new row and column labels.
+
+        Notes
+        -----
+        If level is not specified, the default behavior is to replace row labels in all levels.
+        """
+        pass
+
+    @abstractmethod
+    def sort_by(
+        self, axis: int, columns: Union[str, List[str]], ascending: bool = True
+    ) -> "ModinDataframe":
+        """
+        Logically reorder rows (columns if axis=1) lexicographically by the data in a column or set of columns.
+
+        Parameters
+        ----------
+        axis : int
+            The axis to perform the sort over.
+        columns : string or list of strings
+            Column label(s) to use to determine lexicographical ordering.
+        ascending : boolean, default: True
+            Whether to sort in ascending or descending order.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe sorted into lexicographical order by the specified column(s).
```

Comment:
sort here should be stable because of the importance of dataframe order. Ordering by a custom callback will have to be done with a derived column:

```python
df["order_by_me"] = df["some_col"] + df["other_col"] + something_else  # short hand for mask
df.sort_by("order_by_me")
```